src/main/java/org/assertj/core/internal/ComparatorBasedComparisonStrategy.java

```diff
@@ -94,6 +95,10 @@ public boolean iterableContains(Iterable<?> iterable, Object value) {
   @SuppressWarnings("unchecked")
   public void iterableRemoves(Iterable<?> iterable, Object value) {
     if (iterable == null) return;
+    // Avoid O(N^2) complexity of serial removal from an iterator of collections like ArrayList
+    if (iterable instanceof Collection) {
+      ((Collection<?>) iterable).removeIf(o -> comparator.compare(o, value) == 0);
+    }
     Iterator<?> iterator = iterable.iterator();
     while (iterator.hasNext()) {
```

Comment:
One side note from me. I think that this should be in an `else` statement. Otherwise it would traverse the collection again. This is not really needed, since `removeIf` handles everything. Same for the changes in the `StandardComparisonStrategy`

Reply:
Thanks! Fixed.